kubespawner/spawner.py

```diff
@@ -118,21 +113,25 @@ class KubeSpawner(Spawner):
     spawned by a user will have its own KubeSpawner instance.
     """
 
-    reflectors = {
-        "pods": None,
-        "events": None,
-    }
+    # Reflectors are stored in class variable for performance reasons.
+    # If every pod will start its own reflector, Kubelet will not be happy
+    # with attaching 20k event watchers
+    reflectors = {}
 
     # Characters as defined by safe for DNS
     # Note: '-' is not in safe_chars, as it is being used as escape character
     safe_chars = set(string.ascii_lowercase + string.digits)
 
+    def get_reflector_key(self, kind: str) -> tuple:
```

Comment:
I'm not sure about this implementation.

From the one hand, tests are working properly without calling any additional method to reset reflectors. From another, changes needed by tests are added into spawner class

Reply:
Can you clarify where and why tests needed this change?

Reply:
I've added test `test_spawn_component_label` which starts pod with different `component_label` value. But because reflectors use `component_label` to get events from k8s, and reflector instance is just one per KubeSpawner class, previous tests already set different `component_label`, which causes failure of `test_spawn_component_label`.

There are 3 ways to solve this:
1. Create reflector per spawner object instead of the whole class. Cannot be done because this reduces performance
2. Reset reflector object before and after test with a different `component_label`. I've tried to do this previously, but calling private class method in tests is quite strange for me.
3. Create reflector instance for each component_label (plus namespace, if reflector is watching for a specific namespace)

Now I implemented option 3, is that OK?

Reply:
I think overall KubeSpawner's reflectors are the hardest thing to understand and maintain, so adding complexity in there to support a test makes me uneasy.

I wonder if updating tests to avoid re-using old configuration of reflectors or similar is possible. I'd like to push a bit more on strategy 2 before resorting to strategy 3, I'll see if I can figure something out.

From now on to ease my review process, do not make any force pushes but instead add commits if you make changes.

Reply:
For reference, this is where I'm looking: https://github.com/jupyterhub/kubespawner/pull/654/files#diff-84f25ce0bea68e4626c0cc2f38be2ec945648eddce266b2bbac73519390756f9R209-R222

How and can we get the KubeSpawner class state be reset between these tests?

Reply:
This was in previous implementation I've pushed to this PR:
https://github.com/jupyterhub/kubespawner/commit/89e645c9ddfaf1f268dea0b11d375fd7b6aa4015#diff-6fdb3a928956e8ab2b810679d51fae72870d45ed4a088e85e16814339aff8901R2429
https://github.com/jupyterhub/kubespawner/commit/89e645c9ddfaf1f268dea0b11d375fd7b6aa4015#diff-84f25ce0bea68e4626c0cc2f38be2ec945648eddce266b2bbac73519390756f9R225-R234

Reply:
Thanks!

Okay I've pushed a commit to this PR, reverting a few things related to strategy 3 above in favor of my approach of strategy 2. What do you think?